src/pymmcore_plus/mda/_engine.py

```diff
@@ -54,11 +67,51 @@ def setup_event(self, event: MDAEvent) -> None:
         self._mmc.waitForSystem()
 
     def exec_event(self, event: MDAEvent) -> Any:
-        """Execute an individual event and return the image data."""
-        # TODO: add non-aquisition event-specific logic here later
+        """Execute an individual event."""
+        action = getattr(event, "action", None)
+
+        # execute hardware autofocus
+        if isinstance(action, HardwareAutofocus) and event.z_pos is not None:
+            # get position index
+            p_idx = event.index.get("p", None)
+            # run autofocus and get the new z position
+            new_z = self._execute_autofocus(action)
+            # get the correction to apply to each z position
+            self._z_correction[p_idx] = new_z - event.z_pos
+            return None
+
+        # acquire an image and emit the image data
         self._mmc.snapImage()
         return EventPayload(image=self._mmc.getImage())
 
+    def _execute_autofocus(self, action: HardwareAutofocus) -> float:
+        """Perform the hardware autofocus.
+
+        Returns the new z focus position.
+        """
+        self._mmc.setPosition(
+            action.autofocus_device_name,
+            action.autofocus_motor_offset,
+        )
+        self._mmc.waitForSystem()
+
+        # perform fullFocus 3 times in case of failure
```

Comment:
I think this should be an option on the action actually:
`max_retries: int = 3` ...

Reply:
You mean an option of the `useq.HardwareAutofocus` `Action` like this?

```py
class HardwareAutofocus(Action):
    """
    ...
    max_retries : int
        The number of retries if autofocus fails.
    """

    type: Literal["hardware_autofocus"] = "hardware_autofocus"
    autofocus_device_name: str
    autofocus_motor_offset: float
    max_retries: int
```

and then modify this as:
```py
 # perform fullFocus 3 times in case of failure
 for i in range(action.max_retries):
     try:
         self._mmc.fullFocus()
         self._mmc.waitForSystem()
     except RuntimeError:
         warnings.warn(f"Hardware autofocus failed {i} times.", stacklevel=2)

 return self._mmc.getZPosition()
```

Reply:
yep, exactly, but max_retries can take a default of 3, and you'll need to fix the try loop, cause this one will *always* perform `fullFocus` 3 times, regardless of success or failure, and the the previous one only retried in the case of failure.  It's a fun little challenge: i.e. how can you code a nested try catch with a variable number of retries? :)

Reply:
That's true. This should work:
```py
# perform fullFocus 'action.max_retries' times in case of failure
        for i in range(action.max_retries):
            try:
                self._mmc.fullFocus()
                self._mmc.waitForSystem()
                return self._mmc.getZPosition()

            except RuntimeError:
                warnings.warn(
                    f"Hardware autofocus failed {i + 1} time(s), "
                    f"({action.max_retries - (i + 1)} retries remaining).",
                    stacklevel=2,
                )
        warnings.warn(
            "Hardware autofocus failed. Maximum retries exceeded.", stacklevel=2
        )
        return self._mmc.getZPosition()
```

Reply:
yep!  minor suggestions:
I don't think it needs a warning on each try... so you can just use:
```python
with suppress(RuntimeError):
    ...
```

and then if it fails, rather than `warnings.warn`. it should come from the logger

Reply:
should be ok now!

Reply:
nice!